wgengine/capture/capture.go

```diff
@@ -0,0 +1,150 @@
+	binary.Write(w, binary.LittleEndian, uint32(65535))      // max packet len
+	binary.Write(w, binary.LittleEndian, uint32(147))        // link-layer ID - USER0
+}
+
+func writePktHeader(w *bytes.Buffer, when time.Time, length int) {
+	s := when.Unix()
+	us := when.UnixMicro() - (s * 1000000)
+
+	binary.Write(w, binary.LittleEndian, uint32(s))      // timestamp in seconds
+	binary.Write(w, binary.LittleEndian, uint32(us))     // timestamp microseconds
+	binary.Write(w, binary.LittleEndian, uint32(length)) // length present
+	binary.Write(w, binary.LittleEndian, uint32(length)) // total length
+}
+
+// Path describes where in the data path the packet was captured.
+type Path uint8
+
+// Valid Path values.
+const (
+	// FromLocal indicates the packet was logged as it traversed the FromLocal path:
+	// i.e.: A packet from the local system into the TUN.
+	FromLocal Path = 0
+	// FromPeer indicates the packet was logged upon reception from a remote peer.
+	FromPeer Path = 1
+	// SynthesizedToLocal indicates the packet was generated from within tailscaled,
+	// and is being routed to the local machine's network stack.
+	SynthesizedToLocal Path = 2
+	// SynthesizedToPeer indicates the packet was generated from within tailscaled,
+	// and is being routed to a remote Wireguard peer.
+	SynthesizedToPeer Path = 3
+)
+
+// New creates a new capture sink.
+func New() (*Sink, error) {
+	s := &Sink{
+		close: make(chan struct{}),
+	}
+	return s, nil
+}
+
+// Type Sink handles callbacks with packets to be logged,
+// formatting them into a pcap stream which is mirrored to
+// all registered outputs.
+type Sink struct {
+	close chan struct{}
+
+	mu      sync.Mutex
+	outputs []io.Writer
+}
+
+// RegisterOutput connects an output to this sink, which
+// will be written to with a pcap stream as packets are logged.
+//
+// If w implements io.Closer, it will be closed upon error
+// or when the sink is closed. If w implements http.Flusher,
+// it will be flushed after each packet.
+func (s *Sink) RegisterOutput(w io.Writer) {
+	select {
+	case <-s.close:
+		return
+	default:
+	}
+
+	writePcapHeader(w)
+	s.mu.Lock()
+	s.outputs = append(s.outputs, w)
+	s.mu.Unlock()
+}
+
+// Close shuts down the sink. Future calls to LogPacket
+// are ignored, and any registered output that implements
+// io.Closer is closed.
+func (s *Sink) Close() error {
+	close(s.close)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, o := range s.outputs {
+		if o, ok := o.(io.Closer); ok {
+			o.Close()
+		}
+	}
+	s.outputs = nil
+	return nil
+}
+
+// Wait blocks till the Sink is closed.
+func (s *Sink) Wait() {
+	<-s.close
+}
+
+// LogPacket is called to insert a packet into the capture.
+//
+// This function does not take ownership of the provided data slice.
+func (s *Sink) LogPacket(path Path, when time.Time, data []byte) {
+	select {
+	case <-s.close:
+		return
+	default:
+	}
+
+	b := bytes.NewBuffer(make([]byte, 0, 16+2+len(data)))
```

Comment:
this seems like it'll be pretty garbage-y (lot of GC activity). sync.Pool these buffers?

Reply:
Done, but PTAL because Ive never used Pool before so not sure if I'm doing it correctly.